raysect/core/math/function/function2d/tests/test_base.py

```diff
@@ -237,4 +357,201 @@ def test_pow_3_arguments(self):
+                    msg="Callable not equals Function2D (f1() != f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (ref_value != self.f1)(x, y), 0.0,
+                    msg="Callable not equals Function2D (f1() != f2()) did not return false when it should."
+                )
+                self.assertEqual(
+                    (less_value < self.f1)(x, y), 1.0,
+                    msg="Callable less than Function2D (f1() < f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (greater_value < self.f1)(x, y), 0.0,
+                    msg="Callable less than Function2D (f1() < f2()) did not return false when it should."
+                )
+                self.assertEqual(
+                    (greater_value > self.f1)(x, y), 1.0,
+                    msg="Callable greater than Function2D (f1() > f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (less_value > self.f1)(x, y), 0.0,
+                    msg="Callable greater than Function2D (f1() > f2()) did not return false when it should."
+                )
+                self.assertEqual(
+                    (less_value <= self.f1)(x, y), 1.0,
+                    msg="Callable less equals Function2D (f1() <= f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (ref_value <= self.f1)(x, y), 1.0,
+                    msg="Callable less equals Function2D (f1() <= f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (greater_value <= self.f1)(x, y), 0.0,
+                    msg="Callable less equals Function2D (f1() <= f2()) did not return false when it should."
+                )
+                self.assertEqual(
+                    (greater_value >= self.f1)(x, y), 1.0,
+                    msg="Callable equals Function2D (f1() >= f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (ref_value >= self.f1)(x, y), 1.0,
+                    msg="Callable greater equals Function2D (f1() >= f2()) did not return true when it should."
+                )
+                self.assertEqual(
+                    (less_value >= self.f1)(x, y), 0.0,
+                    msg="Callable equals Function2D (f1() >= f2()) did not return false when it should."
+                )
+
+    def test_richcmp_function_function(self):
+        v = [-1e10, -7, -0.001, 0.0, 0.00003, 10, 2.3e49]
+        for x in v:
+            for y in v:
+                ref_value = self.f1
+                # Without Function2D.__abs__, manually ensure greater and less values are suitable
+                if self.f1(x, y) >= 0:
+                    greater_value = self.f1 + self.f1 + 1
+                    less_value = self.f1 - self.f1 - 1
```

Comment:
isn't this just -1? Is this what was intended?

Reply:
I remember why I did this now. This is `Constant2D(-1)`, not `-1`. The tests are `function __richcmp__ function`, so I need a `Function2D` object here rather than a scalar for the comparisons. The expression comes from a copy/paste/tweak of the function-scalar test, working around the lack of `__abs__`.

I can either add a comment explaining this, or explicitly use `Constant2D(-1)` instead. Or implement `__abs__` and make this PR even bigger. Which would you prefer?

Reply:
Adding abs would be preferable.

Reply:
OK, I've added abs and rewritten the test to use it. If you're happy with the abs implementation, I'll add 1D and 3D versions of `__richcmp__` and `__abs__` and then mark this PR as ready for merging.